File: app/src/main/java/com/imceits/aungtuntun/alephcodeassignment/ui/DetailAdapter.java
package com.imceits.aungtuntun.alephcodeassignment.ui;

import android.app.AlertDialog;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.imceits.aungtuntun.alephcodeassignment.R;
import com.imceits.aungtuntun.alephcodeassignment.data.pojos.ToolsRentDto;
import com.imceits.aungtuntun.alephcodeassignment.databinding.RentItemBinding;

import java.util.ArrayList;
import java.util.List;

public class DetailAdapter extends BaseAdapter<DetailAdapter.DetailViewHolder, ToolsRentDto> {
    private List<ToolsRentDto> dataList = new ArrayList<>();
    private final ReturnedListener listener;
    private Context mContext;
    DetailAdapter(Context context, ReturnedListener listener) {
        this.listener = listener;
        this.mContext = context;
    }

    @Override
    public void setData(List<ToolsRentDto> toolsRents) {
        dataList = new ArrayList<>();
        dataList = toolsRents;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public DetailViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RentItemBinding itemBinding = RentItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new DetailViewHolder(itemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull DetailViewHolder holder, int position) {
        holder.onBind(dataList.get(position));
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    class DetailViewHolder extends RecyclerView.ViewHolder {
        RentItemBinding itemBinding;

        DetailViewHolder(RentItemBinding rentItemBinding) {
            super(rentItemBinding.getRoot());
            itemBinding = rentItemBinding;
        }

        void onBind(ToolsRentDto toolsRent) {
            itemBinding.setRent(toolsRent);
            itemBinding.executePendingBindings();
            itemBinding.cardRentTool.setOnClickListener(v -> createDialog(toolsRent, getAdapterPosition()));
        }
    }

    private void createDialog(ToolsRentDto toolsRent, int position) {
        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(mContext);
        dialogBuilder.setTitle(mContext.getString(R.string.title_return));
        dialogBuilder.setMessage("Are you sure to mark this tool as returned?");
        dialogBuilder.setNegativeButton("Cancel", (dialog, which) -> dialog.dismiss());
        dialogBuilder.setPositiveButton("OK", (dialog, which) -> {
            listener.returned(toolsRent);
            dataList.remove(toolsRent);
            notifyItemRemoved(position);
        });
        AlertDialog dialog = dialogBuilder.create();
        dialog.show();
    }

    public interface ReturnedListener {
        void returned(ToolsRentDto toolsRent);
    }
}

File: app/src/main/java/com/imceits/aungtuntun/toolsbox/ui/DetailAdapter.kt
package com.imceits.aungtuntun.toolsbox.ui

import android.app.AlertDialog
import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.imceits.aungtuntun.toolsbox.R
import com.imceits.aungtuntun.toolsbox.data.views.ToolsRentDto
import com.imceits.aungtuntun.toolsbox.databinding.RentItemBinding
import com.imceits.aungtuntun.toolsbox.utils.BaseAdapter

class DetailAdapter(private val mContext: Context, private val listener: ReturnedListener): BaseAdapter<DetailAdapter.DetailViewHolder, ToolsRentDto> () {
    private var dataList: MutableList<ToolsRentDto> = mutableListOf()

    override fun setData(data: List<ToolsRentDto>) {
        dataList = data as MutableList<ToolsRentDto>
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DetailViewHolder {
        val itemBinding: RentItemBinding = RentItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return DetailViewHolder(itemBinding)
    }

    override fun getItemCount(): Int {
        return dataList.size
    }

    override fun onBindViewHolder(holder: DetailViewHolder, position: Int) {
        holder.onBind(dataList[position])
        holder.itemBinding.cardTool.setOnClickListener {
            createDialog(dataList[position], position)
        }
    }

    class DetailViewHolder(rentItemBinding: RentItemBinding) : RecyclerView.ViewHolder(rentItemBinding.root) {
        val itemBinding: RentItemBinding = rentItemBinding

        fun onBind(toolsRentDto: ToolsRentDto) {
            itemBinding.data = toolsRentDto
            itemBinding.executePendingBindings()
        }
    }

    private fun createDialog(toolsRentDto: ToolsRentDto, position: Int) {
        val builder: AlertDialog.Builder = AlertDialog.Builder(mContext)
        builder.setTitle(mContext.getString(R.string.return_tool))
        builder.setMessage(mContext.getString(R.string.return_message))
        builder.setNegativeButton(mContext.getString(R.string.button_cancel)
        ) { dialog, which -> dialog.dismiss()  }
        builder.setPositiveButton(mContext.getString(R.string.button_ok)
        ) { dialog, which ->
            listener.returned(toolsRentDto)
            dataList.remove(toolsRentDto)
            notifyItemRemoved(position)
        }
        val dialog: AlertDialog = builder.create()
        dialog.show()
    }

    interface ReturnedListener {
        fun returned(toolsRentDto: ToolsRentDto)
    }
}